polkadot/runtime/parachains/src/inclusion/mod.rs

```diff
@@ -1104,6 +1104,24 @@ impl<T: Config> Pallet<T> {
 		})
 	}
 
+	/// Returns all the `CommittedCandidateReceipt` pending availability for the para provided, if
+	/// any.
+	pub(crate) fn candidates_pending_availability(
```

Comment:
Maybe we'd want to provide a parameter which is an enum to return `Last' or `All` depending on needs. These candidates can be pretty big due to messages and code upgrades, so tailoring the return type would be useful. WDYT ?

Reply:
If we want to be as generic as possible we could have an optional count. If the count is None, return all of them.
Anyway, it's a matter of how is this being used in cumulus.

AFAICT, in pov-recovery, we need all candidates.
OTOH, in `handle_empty_block_announce_data`, we only need the output head data of the last candidate, which can be retrieved already using the much lighter `persisted_validation_data` runtime API (which force enacts all candidates before returning the head data, so the result is the same).

@skunert WDYT?

Reply:
The `handle_empty_block_announce_data` of `RequireSecondedInBlockAnnounce` is rather niche because it is only used for the relay chain consensus, AURA parachain nodes don't use it. However, even in that scenario we would be interested in whether the block that was announced was backed by the relay chain. Could for sure happen that a block is announced that was in the backed blocks but not the latest.

For `pov-recovery` and the parent search in lookahead collator and slot-based we also need all of them.

TBH I had not thought about the different return types of them before, in the [collator parent search](https://github.com/paritytech/polkadot-sdk/blob/c3a5488d3f610222cbbeb121f8d75a7cc98177e3/cumulus/client/consensus/common/src/lib.rs#L335-L342) we will need to switch from `persisted_validation_data` to `candidates_pending_availability`.

DQ: In theory we could for example have three candidates on three different para branches right? Or do we have some constraint that the para block backed in a single RCB slot must be dependent on each other? If not, a `Last` parameter does not make so much sense.

Reply:
> DQ: In theory we could for example have three candidates on three different para branches right? Or do we have some constraint that the para block backed in a single RCB slot must be dependent on each other? If not, a Last parameter does not make so much sense.

The relay chain won't back parachain forks, it only backs chains.
Also, for proper elastic scaling to work at full throughput, we need parachains to not produce forks

Reply:
> The relay chain won't back parachain forks, it only backs chains.

Okay, makes sense. So if I still submit three different forks to the relay two will be ignored.

>  Also, for proper elastic scaling to work at full throughput, we need parachains to not produce forks

That makes sense, this was a scenario that was in my head, not something I would actively support in the collator :+1:

Reply:
Ok, so we just keep it simple like it is right now.